### PR TITLE
Issue 2909

### LINKED ISSUE #2909: sp_BlitzCache - Bug in duplicate plan query
**Version of the script**
`@Version = '8.03', @VersionDate = '20210420'`

**What is the current behavior?**
There's a bug in the section of code that starts with `Checking for single use plans and plans with many queries`

Specifically, in this section, we're getting a count of distinct plan handles grouped by query hash, which makes no sense. I believe this is why the >100% numbers were coming back.

```
    SELECT SUM(x.duplicate_plan_handles) AS duplicate_plan_handles
    FROM (
        SELECT COUNT_BIG(DISTINCT plan_handle) AS duplicate_plan_handles
        FROM sys.dm_exec_query_stats qs
            CROSS APPLY sys.dm_exec_plan_attributes(qs.plan_handle) pa
        WHERE pa.attribute = N'dbid'
        GROUP BY qs.query_hash, pa.value
        HAVING COUNT_BIG(DISTINCT plan_handle) > 5
    ) AS x
```

It may also make sense to modify this CTE to get a count from `dm_exec_query_stats` instead of `dm_exec_cached_plans`:

```
    SELECT COUNT_BIG(*) AS total_plans
    FROM sys.dm_exec_cached_plans AS deqs
    WHERE deqs.cacheobjtype = N'Compiled Plan'
```

Since that would match the section of [code immediately above it](https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/blob/dev/sp_BlitzCache.sql#L1582) that looks for plan cache instability. 

**If the current behavior is a bug, please provide the steps to reproduce.**
Code review

**What is the expected behavior?**
Be logically correct

**Which versions of SQL Server and which OS are affected by this issue? Did this work in previous versions of our procedures?**
All.

Just a note, I'm going to leave this open for a few days for feedback. Not sure if I'm missing something obvious, here.

### PR DESCRIPTION
**Pull request by erikdarlingdata**
Hopefully fixes the plan cache count calculations.

Closes #2909

### CODE REVIEW

File: sp_BlitzCache.sql [hunk 1]
```diff
@@ -1600,49 +1600,90 @@ OPTION (RECOMPILE);
...
-            CROSS APPLY sys.dm_exec_plan_attributes(qs.plan_handle) pa
+        LEFT JOIN sys.dm_exec_procedure_stats ps 
+            ON qs.sql_handle = ps.sql_handle
+        CROSS APPLY sys.dm_exec_plan_attributes(qs.plan_handle) pa
         WHERE pa.attribute = N'dbid'
-        GROUP BY qs.query_hash, pa.value
-        HAVING COUNT_BIG(DISTINCT plan_handle) > 5
+        AND   qs.query_plan_hash <> 0x0000000000000000
+        GROUP BY
+		    qs.query_plan_hash,
```
Comment (BrentOzar) [inline sp_BlitzCache.sql:1622]: Grouping by qs.query_plan_hash is breaking the "Many Duplicate Plans" warning. Comment this out, and it works.

**Comment by BrentOzar** [comment]
I think I'm missing something obvious. I'm trying the repro script from #2909, and I'm using the code from this branch, but I'm getting this:
Screen Shot 2021-05-24 at 8 02 32 AM

The "Multiple Plans (10)" is wrong - it's showing the number of plans from the top 10 only - and the duplicate plans warning isn't showing up in the second result set either.

The Multiple Plans part appears to be from this section, which is grabbing the duplicate queries from ##BlitzCacheProcs - I know you fixed this in an earlier build for this issue, so I just gotta go track that fix down and add it back in:

```
UPDATE  ##BlitzCacheProcs
SET     NumberOfDistinctPlans = distinct_plan_count,
        NumberOfPlans = number_of_plans ,
        plan_multiple_plans = CASE WHEN distinct_plan_count < number_of_plans THEN number_of_plans END
FROM (
        SELECT  COUNT(DISTINCT QueryHash) AS distinct_plan_count,
                COUNT(QueryHash) AS number_of_plans,
                QueryHash,
				DatabaseName
        FROM    ##BlitzCacheProcs
		WHERE SPID = @@SPID
        GROUP BY QueryHash,
		         DatabaseName
) AS x
WHERE ##BlitzCacheProcs.QueryHash = x.QueryHash
AND   ##BlitzCacheProcs.DatabaseName = x.DatabaseName
OPTION (RECOMPILE) ;
```

About the "Many Duplicate Plans" warning not working - that's because of the grouping by query_plan_hash around line 1622:

```
        SELECT
		    COUNT_BIG(qs.query_plan_hash) AS duplicate_plan_hashes
        FROM sys.dm_exec_query_stats qs
        LEFT JOIN sys.dm_exec_procedure_stats ps 
            ON qs.sql_handle = ps.sql_handle
        CROSS APPLY sys.dm_exec_plan_attributes(qs.plan_handle) pa
        WHERE pa.attribute = N'dbid'
        AND   qs.query_plan_hash <> 0x0000000000000000
        GROUP BY
            qs.query_plan_hash,
            qs.query_hash,
			ps.object_id,
            pa.value
        HAVING COUNT_BIG(qs.query_plan_hash) > 5
```

If I comment that first group-by out, then it works fine. I did a quickie pull request for that part.

**Comment by erikdarlingdata** [comment]
I don't think you're missing anything. If that's working for you, I'm fine going with it.

**Comment by BrentOzar** [comment]
OK, cool. I'll figure out the "Multiple Plans" part tomorrow, edit the commit, and then merge that. Thanks man!

**Review by BrentOzar — APPROVED** [review]

**Comment by BrentOzar** [comment]
Merging this in for the "Many Duplicate Plans" warning fix, and then I'll work on the "Multiple Plans (X)" over in the other issue. Thanks again for this, man! You rock.

**Pull request merged**